File: ad_pub_sub_cpp/src/pub_tb3.cpp
#include<rclcpp/rclcpp.hpp>
#include<geometry_msgs/msg/twist.hpp>
#include<sensor_msgs/msg/laser_scan.hpp>
#include<chrono>

using namespace std::chrono_literals;

class MoveTb3: public rclcpp::Node
{
    public:
        MoveTb3(): Node("MoveNode")
        {
            pub_ = this->create_publisher<geometry_msgs::msg::Twist>("/cmd_vel" , 10);
            sub_ = this->create_subscription<sensor_msgs::msg::LaserScan>("/scan" , 10,std::bind(&MoveTb3::ScanCallback , this , std::placeholders::_1));
            timer_ = this->create_wall_timer(500ms , std::bind(&MoveTb3::TimerCallback , this));
        }
    private:
        void ScanCallback(const sensor_msgs::msg::LaserScan::SharedPtr msg)
        {
            RCLCPP_INFO(this->get_logger() , "[0:%f , pi/2: %f ,pi: %f , 3pi/2: %f]" , msg->ranges[0] , msg->ranges[90] , msg->ranges[180] , msg->ranges[270]);
            laser_data = *msg;
        }

        void TimerCallback()
        {
            if (!laser_data.ranges.empty())
            {
                if (laser_data.ranges[0]>0.5)
                {
                    vel.linear.x = 0.1;
                    pub_->publish(vel);
                }
                else
                {
                    vel.linear.x = 0.0;
                    pub_->publish(vel);
                }
            }
        }

        geometry_msgs::msg::Twist vel;
        sensor_msgs::msg::LaserScan laser_data;
        rclcpp::TimerBase::SharedPtr timer_;
        rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr pub_;
        rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr sub_;
};




int main(int argc , char* argv[])
{
    rclcpp::init(argc,argv);
    auto node = std::make_shared<MoveTb3>();
    rclcpp::spin(node);
    rclcpp::shutdown();

    return 0;

}

File: ad_pub_sub_py/ad_pub_sub_py/pub_tb3.py
import rclpy
# import the ROS2 python libraries
from rclpy.node import Node
# import the Twist module from geometry_msgs interface
from geometry_msgs.msg import Twist
# import the LaserScan module from sensor_msgs interface
from sensor_msgs.msg import LaserScan
from rclpy.qos import ReliabilityPolicy , QoSProfile

class MoveTb3(Node):
    def __init__(self):
        super().__init__("move")
        self.pub_ = self.create_publisher(Twist ,"cmd_vel" ,10)
        self.sub_ = self.create_subscription(LaserScan , "scan" , self.ScanCallback,  QoSProfile(reliability = ReliabilityPolicy.BEST_EFFORT , depth = 10) )
        self.sub_
        self.laserData = LaserScan()
        self.create_timer(0.1 , self.timer_callback)

    def ScanCallback(self ,msg):
        self.get_logger().info("many?? %i" % len(msg.ranges))
        # self.get_logger().info("ranges[0] , ranges[pi/2] , ranges [pi] , ranges[3pi/2] , [%s , %s , %s , %s]" %(str(msg.ranges[0]) , str(msg.ranges[360//4]) , str(msg.ranges[360//2]) , str(msg.ranges[270]))) 
        self.laserData = msg

    def timer_callback(self):
        if len(self.laserData.ranges) != 0:
            self.get_logger().info("ranges[0] , ranges[pi/2] , ranges [pi] , ranges[3pi/2] , [%s , %s , %s , %s]" %(str(self.laserData.ranges[0]) , str(self.laserData.ranges[360//4]) , str(self.laserData.ranges[360//2]) , str(self.laserData.ranges[270]))) 
            msg = Twist()
        
            if self.laserData.ranges[0]>0.5 :
                msg.linear.x = 0.1
                self.pub_.publish(msg)
            else:
                msg.linear.x = 0.0
                self.pub_.publish(msg)

       






def main(args=None):
    rclpy.init(args=args)
    node = MoveTb3()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__=='__main__':
    main()